### ISSUE TITLE
Win32 handles may be held

### ISSUE THREAD

**Issue by bdbaddog** [SCons/scons#1721]
This issue was originally created at: 2007-08-30 14:02:36.
This issue was reported by: `ses4j`.
ses4j said at 2007-08-30 14:02:36
>Running SCons 0.97, Win32, Python 2.4.4.
> 
>When running SCons on Win32 with multiple threads (-j2 or higher), we 
>experienced problems with releasing file handles, causing builds to fail 
>occasionally.  Relaunching scons would succeesfully build, however.
> 
>Upon researching, what we found was that scons uses Python's spawnve() to 
>launch build actions, which on Win32 runs CreateProcess with "inheritHandles" 
>on.  This means that any spawned subprocesses (like cmd.exe and cl.exe) inherit 
>the file handles of the main process (presumably python.exe).
> 
>This can cause problems like we ran into.
> 
>For example, assume three Actions are required:
> 
>1) GenerateHeader(target=['MyFile.h'], source=[]) is a Python function that 
>creates and writes "MyHeader.h".
>2) Program('MyProgram', ['MyClass.cpp']) compiles MyClass.cpp which depends on 
>MyFile.h
>3) Program('MyUnrelatedProgram', ['UnrelatedFile.cpp']) does -not- depend on 
>any other files.
> 
>- Running "scons -j2" launches actions 1 and 3 in parallel.  
>- However, action 1 was launched first, and quickly got a write lock on 
>MyFile.h.  
>- Action 1 was still processing while Action 3 launched with CreateProcess
>(inheritHandles=true), which means that Action 3 inherited the write lock on 
>MyFile.h.
>- Action 1 completes and unlocks MyFile.h.  Action 3 is still going, and 
>retains the lock on MyFile.h.
>- Action 2 begins, and cl.exe demands an exclusive lock on MyFile.h.  It can't 
>get it since Action 3's process still has it's inherited write lock on 
>MyFile.h.  The build fails.
>- On retry, all is well since MyFile.h is already generated.  Nothing ever gets 
>the write lock, and Action 2 reads it just fine.
> 
>Recommendations / Workarounds:
> 
>I believe the answer is to not call Python spawnve() on Windows.  Instead, if 
>win32api.CreateProcess is called directly with inheritHandles off, it seems to 
>work okay.  This is where I get fuzzy though: there seem to be some issues with 
>inheritHandles off and using the parent's in/out/err pipes.  I didn't actually 
>SEE problems, but some documentation warnings gave me pause.  
> 
>It can be worked around as-is by avoiding locking files in Python Actions that 
>will be read by cl.exe or other things that need an exclusive lock.

ses4j said at 2007-09-05 13:13:34
>I believe this same issue impacts many more people on Windows than I thought.  
>I think any cache retrieve has the potential to cause the same error on build.
> 
>For example, if one build thread does a cache retrieve (including the python 
>function copy2), it acquires an exclusive write lock on the target file.  It 
>lets go of that lock, but not before build thread 2, during the copy, spawns a 
>process which inherits the lock.  Thread 1 completes the CacheRetrieve and 
>begins a new action, which requires a read lock on that file and fails.
> 
>This problem is causing our project significant performance loss, since builds 
>must be run either -j1 or run multiple times.

dsvensson said at 2007-12-04 05:24:40
>More information on this bug can be found in this thread:
> 
>http://article.gmane.org/gmane.comp.programming.tools.scons.user/15414
> 
>This bug is probably a duplicate of:
> 
>http://scons.tigris.org/issues/show_bug.cgi?id=1177

garyo said at 2007-12-04 05:55:58
>*** Issue 1177 has been marked as a duplicate of this issue. ***

garyo said at 2007-12-04 05:56:58
>See issue 1721 for a test case that fails (python-generated source file fed
>immediately to compiler).

garyo said at 2007-12-04 05:58:38
>sorry, I meant issue 1177 (see comment above).
>http://scons.tigris.org/issues/show_bug.cgi?id=1177

stevenknight said at 2007-12-04 07:02:34
>Created an attachment (id=258)
>test case from Issue 1177

ses4j said at 2007-12-04 17:16:29
>Our team has come up with this possible plan for a patch, though we haven't as
>yet had a chance to implement anything useful:
> 
>When scons -jX launches on Win32, spawn X worker processes.  Each worker process
>can set up a pipe to the main process with which to communicate jobs (command
>line and environment) and return return codes.  When scons goes to actually
>spawn a job, rather, it will just pipe orders to one of the worker processes and
>block until the return code comes back.
> 
>This should circumvent the issue since by launching those worker processes early
>on in the scons run, no evil handles will be inherited by the subprocesses.  It
>may also show a performance gain, as there's no need to call CreateProcess so
>many times.

garyo said at 2007-12-05 06:01:28
>This is clever, Scott, but I'm not sure it would actually be faster. 
>Marshalling the command and env and sending it over a pipe (and then the worker
>job waking up and running the command) might not win over CreateProcess.  I
>think the worker job would end up running CreateProcess anyway (?).  It'll be a
>nontrivial amount of work to implement well, I think.  It certainly should be
>safe enough, but if the existing patch can be shown to not inherit handles while
>preserving the ability to get stdio and status codes back to scons, it seems
>much more straightforward.

stevenknight said at 2007-12-08 10:21:10
>Re: Scott's suggestion:
> 
>Gary's pretty much on target here.  The suggestions that this is a threading
>issue (particularly my descriptions of it) have been somewhat misleading.  It's
>really a general process-creation issue.
> 
>In fact, SCons already has a thread architecture like what SCott describes.  We
>spawn N worker threads to begin with and send builds requests to each on a
>queue.  But they're not separate processes.  And we'd still have to
>CreateProcess() to do things like start up the compiler, that's just not going
>to be handled in a native thread.
> 
>This discussion that I stumbled on in the Python developers mailing list:
> 
>    http://mail.python.org/pipermail/python-dev/2007-June/073699.html
> 
>Also reinforces the idea that this is a general issue with how Python deals with
>Windows file handles across process creation, not something specific to SCons'
>thread architecture.
> 
>Nevertheless, many thanks for the suggestion -- keep the ideas coming!
> 
>After submitting this message, I'm going ahead with checking in a fix that
>replaces __builtin__.close() and __builtin__.file() with wrappers that disable
>the inherit-the-handle flag on the newly opened files, which seems to fix the
>problem while still allow the subprocesses access to the standard input,
>standard output and error output handles.
> 
>Thanks,

stevenknight said at 2007-12-08 11:44:14
>Should be fixed by the following check in, which makes sure that file handles
>for all files opened during the SCons run are set by default to be not inherited
>by subprocesses:
> 
>http://scons.tigris.org/servlets/ReadMsg?list=dev&msgNo=4533
>http://scons.tigris.org/servlets/ReadMsg?list=commits&msgNo=1084
> 
>Thanks,

Votes for this issue: 18.

stevenknight attached [parallel.zip](https://github.com/SCons/tigris-issue-attachments/blob/master/1721/258/parallel.zip) at 2008-03-25 11:08:15.
>test case from Issue 1177

garyo said this issue is duplicated by #1177 at 2007-12-04 05:55:59.
